File: native/core/src/edit/edit_spec.cpp
// SPDX-License-Identifier: Apache-2.0
// Copyright (c) 2026 Pablo contributors.

#include "edit/edit_spec.h"

#include <array>
#include <cmath>
#include <cstdlib>
#include <cstdio>
#include <cstring>
#include <string_view>

namespace photo::edit {

namespace {

constexpr double kEps = 1e-6;

bool nearly_zero(double v) { return std::fabs(v) < kEps; }

// strtod the [begin,end) view (NUL-terminated copy via std::string is avoided on
// the hot path by using a small stack buffer; values are short).
double to_double(std::string_view v, double fallback) {
    if (v.empty()) return fallback;
    std::array<char, 64> buf{};
    const size_t n = v.size() < buf.size() - 1 ? v.size() : buf.size() - 1;
    std::memcpy(buf.data(), v.data(), n);
    buf[n] = '\0';
    char* endp = nullptr;
    const double d = std::strtod(buf.data(), &endp);
    // Reject empty parses, trailing garbage, and non-finite values (strtod
    // happily parses "inf"/"nan", which would poison the render math).
    if (endp == buf.data() || *endp != '\0' || !std::isfinite(d)) return fallback;
    return d;
}

int to_int(std::string_view v, int fallback) {
    return static_cast<int>(to_double(v, static_cast<double>(fallback)));
}

bool to_bool(std::string_view v) {
    return v == "1" || v == "true";
}

// Format a double compactly: integers print without a decimal point, others
// with up to 4 significant fractional digits and trailing zeros trimmed.
std::string fmt(double v) {
    if (nearly_zero(v - std::round(v)) && std::fabs(v) < 1e15) {
        char b[32];
        std::snprintf(b, sizeof(b), "%lld", static_cast<long long>(std::llround(v)));
        return b;
    }
    // Big enough for any finite double's "%.4f" form (~316 chars at 1e308) so a
    // large-but-finite value serializes faithfully rather than truncating.
    char b[512];
    std::snprintf(b, sizeof(b), "%.4f", v);
    std::string s(b);
    // Trim trailing zeros, then a dangling '.'.
    size_t last = s.find_last_not_of('0');
    if (last != std::string::npos) {
        if (s[last] == '.') --last;
        s.erase(last + 1);
    }
    return s;
}

void append_kv(std::string& out, const char* key, const std::string& val) {
    out += key;
    out += '=';
    out += val;
    out += ';';
}

// Percent-encode the grammar delimiters (and control chars) so arbitrary text
// can live inside a spec value. Reversed by unesc().
std::string esc(const std::string& s) {
    std::string o;
    o.reserve(s.size());
    for (unsigned char c : s) {
        if (c == '%' || c == ';' || c == '=' || c == ',' || c == '|' || c < 0x20) {
            char b[4];
            std::snprintf(b, sizeof(b), "%%%02X", c);
            o += b;
        } else {
            o += static_cast<char>(c);
        }
    }
    return o;
}

// Parse a `x,y,r|x,y,r` region list (shared by redeye + heal).
std::vector<Region> parse_regions(std::string_view v) {
    std::vector<Region> out;
    size_t p = 0;
    while (p <= v.size()) {
        size_t bar = v.find('|', p);
        if (bar == std::string_view::npos) bar = v.size();
        std::string_view it = v.substr(p, bar - p);
        p = bar + 1;
        if (it.empty()) continue;
        const size_t c1 = it.find(',');
        const size_t c2 = c1 == std::string_view::npos ? c1 : it.find(',', c1 + 1);
        if (c2 == std::string_view::npos) continue;
        Region r;
        r.x = static_cast<float>(to_double(it.substr(0, c1), 0.5));
        r.y = static_cast<float>(to_double(it.substr(c1 + 1, c2 - c1 - 1), 0.5));
        r.r = static_cast<float>(to_double(it.substr(c2 + 1), 0.04));
        out.push_back(r);
    }
    return out;
}

std::string serialize_regions(const std::vector<Region>& rs) {
    std::string out;
    for (size_t i = 0; i < rs.size(); ++i) {
        if (i) out += '|';
        out += fmt(rs[i].x) + "," + fmt(rs[i].y) + "," + fmt(rs[i].r);
    }
    return out;
}

std::string unesc(std::string_view s) {
    auto hex = [](char c) -> int {
        if (c >= '0' && c <= '9') return c - '0';
        if (c >= 'A' && c <= 'F') return c - 'A' + 10;
        if (c >= 'a' && c <= 'f') return c - 'a' + 10;
        return -1;
    };
    std::string o;
    o.reserve(s.size());
    for (size_t i = 0; i < s.size(); ++i) {
        if (s[i] == '%' && i + 2 < s.size()) {
            const int h = hex(s[i + 1]), l = hex(s[i + 2]);
            if (h >= 0 && l >= 0) {
                o += static_cast<char>(h * 16 + l);
                i += 2;
                continue;
            }
        }
        o += s[i];
    }
    return o;
}

}  // namespace

bool EditSpec::has_geometry() const {
    return rot90 != 0 || flipH || flipV || !nearly_zero(straighten) ||
           !nearly_zero(cropL) || !nearly_zero(cropT) ||
           !nearly_zero(cropW - 1.0) || !nearly_zero(cropH - 1.0) ||
           cropW <= 0.0;
}

bool EditSpec::curve_is_identity() const {
    if (curve.empty()) return true;
    for (const auto& [x, y] : curve)
        if (std::fabs(x - y) > 1e-3) return false;
    return true;
}

bool EditSpec::has_tone_ops() const {
    if (autoFix) return true;
    if (!curve_is_identity()) return true;
    const bool filter_default = filter.empty() || filter == "none";
    return !(filter_default && nearly_zero(exposure) && nearly_zero(contrast) &&
             nearly_zero(highlights) && nearly_zero(shadows) &&
             nearly_zero(whites) && nearly_zero(blacks) && nearly_zero(clarity) &&
             nearly_zero(dehaze) && nearly_zero(temperature) &&
             nearly_zero(tint) && nearly_zero(vibrance) &&
             nearly_zero(saturation) && nearly_zero(sharpness) &&
             nearly_zero(noise) && nearly_zero(vignette));
}

bool EditSpec::has_pixel_ops() const {
    return has_tone_ops() || !redeye.empty() || !heal.empty() || !texts.empty();
}

bool EditSpec::is_identity() const {
    return !has_geometry() && !has_pixel_ops();
}

EditSpec parse_edit_spec(const std::string& s) {
    EditSpec e;
    size_t i = 0;
    const size_t n = s.size();
    while (i < n) {
        size_t semi = s.find(';', i);
        if (semi == std::string::npos) semi = n;
        std::string_view tok(s.data() + i, semi - i);
        i = semi + 1;
        if (tok.empty()) continue;
        const size_t eq = tok.find('=');
        if (eq == std::string_view::npos) continue;
        const std::string_view k = tok.substr(0, eq);
        const std::string_view v = tok.substr(eq + 1);

        if (k == "rot") e.rot90 = ((to_int(v, 0) % 4) + 4) % 4;
        else if (k == "fliph") e.flipH = to_bool(v);
        else if (k == "flipv") e.flipV = to_bool(v);
        else if (k == "straighten") e.straighten = to_double(v, 0);
        else if (k == "crop") {
            // value is "l,t,w,h"
            double c[4] = {0, 0, 1, 1};
            size_t p = 0;
            for (int idx = 0; idx < 4 && p <= v.size(); ++idx) {
                size_t comma = v.find(',', p);
                if (comma == std::string_view::npos) comma = v.size();
                c[idx] = to_double(v.substr(p, comma - p), c[idx]);
                p = comma + 1;
            }
            e.cropL = c[0]; e.cropT = c[1]; e.cropW = c[2]; e.cropH = c[3];
        }
        else if (k == "exposure") e.exposure = to_double(v, 0);
        else if (k == "contrast") e.contrast = to_double(v, 0);
        else if (k == "highlights") e.highlights = to_double(v, 0);
        else if (k == "shadows") e.shadows = to_double(v, 0);
        else if (k == "whites") e.whites = to_double(v, 0);
        else if (k == "blacks") e.blacks = to_double(v, 0);
        else if (k == "clarity") e.clarity = to_double(v, 0);
        else if (k == "dehaze") e.dehaze = to_double(v, 0);
        else if (k == "temp") e.temperature = to_double(v, 0);
        else if (k == "tint") e.tint = to_double(v, 0);
        else if (k == "vibrance") e.vibrance = to_double(v, 0);
        else if (k == "saturation") e.saturation = to_double(v, 0);
        else if (k == "sharpness") e.sharpness = to_double(v, 0);
        else if (k == "noise") e.noise = to_double(v, 0);
        else if (k == "vignette") e.vignette = to_double(v, 0);
        else if (k == "autofix") e.autoFix = to_bool(v);
        else if (k == "curves") {
            e.curve.clear();
            size_t p = 0;
            while (p <= v.size()) {
                size_t bar = v.find('|', p);
                if (bar == std::string_view::npos) bar = v.size();
                std::string_view pt = v.substr(p, bar - p);
                const size_t comma = pt.find(',');
                if (comma != std::string_view::npos)
                    e.curve.emplace_back(
                        static_cast<float>(to_double(pt.substr(0, comma), 0)),
                        static_cast<float>(to_double(pt.substr(comma + 1), 0)));
                p = bar + 1;
            }
        }
        else if (k == "text") {
            e.texts.clear();
            size_t p = 0;
            while (p <= v.size()) {
                size_t bar = v.find('|', p);
                if (bar == std::string_view::npos) bar = v.size();
                std::string_view it = v.substr(p, bar - p);
                p = bar + 1;
                if (it.empty()) continue;
                const size_t c1 = it.find(',');
                const size_t c2 = c1 == std::string_view::npos ? c1 : it.find(',', c1 + 1);
                const size_t c3 = c2 == std::string_view::npos ? c2 : it.find(',', c2 + 1);
                const size_t c4 = c3 == std::string_view::npos ? c3 : it.find(',', c3 + 1);
                if (c4 == std::string_view::npos) continue;
                TextItem t;
                t.x = static_cast<float>(to_double(it.substr(0, c1), 0.5));
                t.y = static_cast<float>(to_double(it.substr(c1 + 1, c2 - c1 - 1), 0.5));
                t.size = static_cast<float>(to_double(it.substr(c2 + 1, c3 - c2 - 1), 0.06));
                t.color = static_cast<uint32_t>(std::strtoul(
                    std::string(it.substr(c3 + 1, c4 - c3 - 1)).c_str(), nullptr, 16));
                t.text = unesc(it.substr(c4 + 1));
                e.texts.push_back(std::move(t));
            }
        }
        else if (k == "redeye") e.redeye = parse_regions(v);
        else if (k == "heal") e.heal = parse_regions(v);
        else if (k == "filter") e.filter = std::string(v);
        // unknown keys ignored (forward-compat)
    }
    return e;
}

std::string serialize_edit_spec(const EditSpec& e) {
    std::string out;
    // Geometry
    if (e.rot90 != 0) append_kv(out, "rot", fmt(e.rot90));
    if (e.flipH) append_kv(out, "fliph", "1");
    if (e.flipV) append_kv(out, "flipv", "1");
    if (!nearly_zero(e.straighten)) append_kv(out, "straighten", fmt(e.straighten));
    if (e.has_geometry() &&
        (!nearly_zero(e.cropL) || !nearly_zero(e.cropT) ||
         !nearly_zero(e.cropW - 1.0) || !nearly_zero(e.cropH - 1.0))) {
        append_kv(out, "crop",
                  fmt(e.cropL) + "," + fmt(e.cropT) + "," +
                  fmt(e.cropW) + "," + fmt(e.cropH));
    }
    // Tone
    if (!nearly_zero(e.exposure)) append_kv(out, "exposure", fmt(e.exposure));
    if (!nearly_zero(e.contrast)) append_kv(out, "contrast", fmt(e.contrast));
    if (!nearly_zero(e.highlights)) append_kv(out, "highlights", fmt(e.highlights));
    if (!nearly_zero(e.shadows)) append_kv(out, "shadows", fmt(e.shadows));
    if (!nearly_zero(e.whites)) append_kv(out, "whites", fmt(e.whites));
    if (!nearly_zero(e.blacks)) append_kv(out, "blacks", fmt(e.blacks));
    if (!nearly_zero(e.clarity)) append_kv(out, "clarity", fmt(e.clarity));
    if (!nearly_zero(e.dehaze)) append_kv(out, "dehaze", fmt(e.dehaze));
    // Colour
    if (!nearly_zero(e.temperature)) append_kv(out, "temp", fmt(e.temperature));
    if (!nearly_zero(e.tint)) append_kv(out, "tint", fmt(e.tint));
    if (!nearly_zero(e.vibrance)) append_kv(out, "vibrance", fmt(e.vibrance));
    if (!nearly_zero(e.saturation)) append_kv(out, "saturation", fmt(e.saturation));
    // Detail
    if (!nearly_zero(e.sharpness)) append_kv(out, "sharpness", fmt(e.sharpness));
    if (!nearly_zero(e.noise)) append_kv(out, "noise", fmt(e.noise));
    if (!nearly_zero(e.vignette)) append_kv(out, "vignette", fmt(e.vignette));
    if (e.autoFix) append_kv(out, "autofix", "1");
    // Curves
    if (!e.curve_is_identity()) {
        std::string val;
        for (size_t i = 0; i < e.curve.size(); ++i) {
            if (i) val += '|';
            val += fmt(e.curve[i].first) + "," + fmt(e.curve[i].second);
        }
        append_kv(out, "curves", val);
    }
    // Text overlays
    if (!e.texts.empty()) {
        std::string val;
        for (size_t i = 0; i < e.texts.size(); ++i) {
            if (i) val += '|';
            const TextItem& t = e.texts[i];
            char col[8];
            std::snprintf(col, sizeof(col), "%06X", t.color & 0xFFFFFFu);
            val += fmt(t.x) + "," + fmt(t.y) + "," + fmt(t.size) + "," + col +
                   "," + esc(t.text);
        }
        append_kv(out, "text", val);
    }
    // Retouch (post-geometry, normalized coords)
    if (!e.redeye.empty()) append_kv(out, "redeye", serialize_regions(e.redeye));
    if (!e.heal.empty()) append_kv(out, "heal", serialize_regions(e.heal));
    // Filter
    if (!e.filter.empty() && e.filter != "none") append_kv(out, "filter", e.filter);
    return out;
}

}  // namespace photo::edit

File: pablo/lib/features/editor/edit_spec.dart
// EditSpec — the Dart mirror of the native parametric edit stack
// (native/core/src/edit/edit_spec). Serializes to the same compact
// `key=value;` grammar so a spec authored here renders identically when sent to
// `previewEdits` / `setAssetEdits`. Only non-default fields are emitted, so an
// identity spec encodes to "".

import 'dart:math' as math;
import 'dart:ui' show Offset;

/// A circular retouch region (mirrors native Region): normalized centre plus a
/// radius expressed as a fraction of the image's SHORT edge. Used for both
/// red-eye circles and heal brush dabs.
class EditRegion {
  EditRegion({this.x = 0.5, this.y = 0.5, this.r = 0.04});
  double x, y, r;

  EditRegion clone() => EditRegion(x: x, y: y, r: r);
}

/// One text overlay (mirrors native TextItem). Position + size are normalized to
/// the image; size is a fraction of image height; color is 0xRRGGBB.
class TextOverlay {
  TextOverlay({
    this.x = 0.5,
    this.y = 0.5,
    this.size = 0.06,
    this.color = 0xFFFFFF,
    this.text = '',
  });
  double x, y, size;
  int color;
  String text;

  TextOverlay clone() =>
      TextOverlay(x: x, y: y, size: size, color: color, text: text);
}

class EditSpec {
  EditSpec({
    this.rot90 = 0,
    this.flipH = false,
    this.flipV = false,
    this.straighten = 0,
    this.cropL = 0,
    this.cropT = 0,
    this.cropW = 1,
    this.cropH = 1,
    this.exposure = 0,
    this.contrast = 0,
    this.highlights = 0,
    this.shadows = 0,
    this.whites = 0,
    this.blacks = 0,
    this.clarity = 0,
    this.dehaze = 0,
    this.temperature = 0,
    this.tint = 0,
    this.vibrance = 0,
    this.saturation = 0,
    this.sharpness = 0,
    this.noise = 0,
    this.vignette = 0,
    this.autoFix = false,
    List<Offset>? curve,
    List<EditRegion>? redeye,
    List<EditRegion>? heal,
    List<TextOverlay>? texts,
    this.filter = 'none',
  })  : curve = curve ?? <Offset>[],
        redeye = redeye ?? <EditRegion>[],
        heal = heal ?? <EditRegion>[],
        texts = texts ?? <TextOverlay>[];

  // Geometry.
  int rot90;
  bool flipH;
  bool flipV;
  double straighten;
  double cropL, cropT, cropW, cropH;

  // Tone (Light).
  double exposure, contrast, highlights, shadows, whites, blacks, clarity, dehaze;

  // Colour.
  double temperature, tint, vibrance, saturation;

  // Detail.
  double sharpness, noise, vignette;

  // One-click enhance (auto-levels).
  bool autoFix;

  // Master tone curve control points (normalized); empty/diagonal = identity.
  List<Offset> curve;

  // Retouch (post-geometry, normalized coords).
  List<EditRegion> redeye; // red-eye correction circles
  List<EditRegion> heal; // heal / spot-removal dabs

  // Text overlays (rendered last, on top).
  List<TextOverlay> texts;

  // Filter preset id (matches kEditorFilters / filter_matrices.dart).
  String filter;

  bool get curveIsIdentity {
    if (curve.isEmpty) return true;
    for (final p in curve) {
      if ((p.dx - p.dy).abs() > 1e-3) return false;
    }
    return true;
  }

  static const double _eps = 1e-6;
  static bool _z(double v) => v.abs() < _eps;

  bool get hasGeometry =>
      rot90 != 0 ||
      flipH ||
      flipV ||
      !_z(straighten) ||
      !_z(cropL) ||
      !_z(cropT) ||
      !_z(cropW - 1) ||
      !_z(cropH - 1) ||
      cropW <= 0;

  bool get isIdentity =>
      !hasGeometry &&
      !autoFix &&
      curveIsIdentity &&
      redeye.isEmpty &&
      heal.isEmpty &&
      texts.isEmpty &&
      (filter.isEmpty || filter == 'none') &&
      _z(exposure) &&
      _z(contrast) &&
      _z(highlights) &&
      _z(shadows) &&
      _z(whites) &&
      _z(blacks) &&
      _z(clarity) &&
      _z(dehaze) &&
      _z(temperature) &&
      _z(tint) &&
      _z(vibrance) &&
      _z(saturation) &&
      _z(sharpness) &&
      _z(noise) &&
      _z(vignette);

  EditSpec clone() => EditSpec(
        rot90: rot90,
        flipH: flipH,
        flipV: flipV,
        straighten: straighten,
        cropL: cropL,
        cropT: cropT,
        cropW: cropW,
        cropH: cropH,
        exposure: exposure,
        contrast: contrast,
        highlights: highlights,
        shadows: shadows,
        whites: whites,
        blacks: blacks,
        clarity: clarity,
        dehaze: dehaze,
        temperature: temperature,
        tint: tint,
        vibrance: vibrance,
        saturation: saturation,
        sharpness: sharpness,
        noise: noise,
        vignette: vignette,
        autoFix: autoFix,
        curve: List<Offset>.from(curve),
        redeye: redeye.map((r) => r.clone()).toList(),
        heal: heal.map((r) => r.clone()).toList(),
        texts: texts.map((t) => t.clone()).toList(),
        filter: filter,
      );

  /// Reset every field to its neutral default in place.
  void reset() {
    rot90 = 0;
    flipH = flipV = false;
    straighten = 0;
    cropL = cropT = 0;
    cropW = cropH = 1;
    exposure = contrast = highlights = shadows = whites = blacks = 0;
    clarity = dehaze = temperature = tint = vibrance = saturation = 0;
    sharpness = noise = vignette = 0;
    autoFix = false;
    curve = <Offset>[];
    redeye = <EditRegion>[];
    heal = <EditRegion>[];
    texts = <TextOverlay>[];
    filter = 'none';
  }

  static String _fmt(double v) {
    if ((v - v.roundToDouble()).abs() < _eps) {
      return v.round().toString();
    }
    var s = v.toStringAsFixed(4);
    // Trim trailing zeros then a dangling dot.
    s = s.replaceFirst(RegExp(r'0+$'), '');
    if (s.endsWith('.')) s = s.substring(0, s.length - 1);
    return s;
  }

  // Percent-encode the grammar delimiters so arbitrary text survives a spec
  // value (mirrors native esc/unesc in edit_spec.cpp).
  static String _esc(String s) {
    final b = StringBuffer();
    for (final cu in s.runes) {
      if (cu == 0x25 || cu == 0x3B || cu == 0x3D || cu == 0x2C || cu == 0x7C ||
          cu < 0x20) {
        b.write('%${cu.toRadixString(16).toUpperCase().padLeft(2, '0')}');
      } else {
        b.writeCharCode(cu);
      }
    }
    return b.toString();
  }

  // Parse a numeric field, mirroring native to_double: reject non-finite
  // (Dart's double.tryParse admits "Infinity"/"NaN" where strtod+isfinite in
  // edit_spec.cpp rejects them) and any unparseable token, falling back to [fb].
  static double _num(String v, double fb) {
    final x = double.tryParse(v);
    return (x != null && x.isFinite) ? x : fb;
  }

  static String _unesc(String s) {
    final b = StringBuffer();
    for (var i = 0; i < s.length; i++) {
      if (s[i] == '%' && i + 2 < s.length) {
        final h = int.tryParse(s.substring(i + 1, i + 3), radix: 16);
        if (h != null) {
          b.writeCharCode(h);
          i += 2;
          continue;
        }
      }
      b.write(s[i]);
    }
    return b.toString();
  }

  /// Serialize to the compact `key=value;` grammar (only non-default fields).
  String encode() {
    final b = StringBuffer();
    void kv(String k, String v) {
      b
        ..write(k)
        ..write('=')
        ..write(v)
        ..write(';');
    }

    if (rot90 != 0) kv('rot', '$rot90');
    if (flipH) kv('fliph', '1');
    if (flipV) kv('flipv', '1');
    if (!_z(straighten)) kv('straighten', _fmt(straighten));
    if (hasGeometry &&
        (!_z(cropL) || !_z(cropT) || !_z(cropW - 1) || !_z(cropH - 1))) {
      kv('crop',
          '${_fmt(cropL)},${_fmt(cropT)},${_fmt(cropW)},${_fmt(cropH)}');
    }
    if (!_z(exposure)) kv('exposure', _fmt(exposure));
    if (!_z(contrast)) kv('contrast', _fmt(contrast));
    if (!_z(highlights)) kv('highlights', _fmt(highlights));
    if (!_z(shadows)) kv('shadows', _fmt(shadows));
    if (!_z(whites)) kv('whites', _fmt(whites));
    if (!_z(blacks)) kv('blacks', _fmt(blacks));
    if (!_z(clarity)) kv('clarity', _fmt(clarity));
    if (!_z(dehaze)) kv('dehaze', _fmt(dehaze));
    if (!_z(temperature)) kv('temp', _fmt(temperature));
    if (!_z(tint)) kv('tint', _fmt(tint));
    if (!_z(vibrance)) kv('vibrance', _fmt(vibrance));
    if (!_z(saturation)) kv('saturation', _fmt(saturation));
    if (!_z(sharpness)) kv('sharpness', _fmt(sharpness));
    if (!_z(noise)) kv('noise', _fmt(noise));
    if (!_z(vignette)) kv('vignette', _fmt(vignette));
    if (autoFix) kv('autofix', '1');
    if (!curveIsIdentity) {
      kv('curves',
          curve.map((p) => '${_fmt(p.dx)},${_fmt(p.dy)}').join('|'));
    }
    if (texts.isNotEmpty) {
      kv(
        'text',
        texts.map((t) {
          final col =
              (t.color & 0xFFFFFF).toRadixString(16).toUpperCase().padLeft(6, '0');
          return '${_fmt(t.x)},${_fmt(t.y)},${_fmt(t.size)},$col,${_esc(t.text)}';
        }).join('|'),
      );
    }
    if (redeye.isNotEmpty) {
      kv('redeye',
          redeye.map((r) => '${_fmt(r.x)},${_fmt(r.y)},${_fmt(r.r)}').join('|'));
    }
    if (heal.isNotEmpty) {
      kv('heal',
          heal.map((r) => '${_fmt(r.x)},${_fmt(r.y)},${_fmt(r.r)}').join('|'));
    }
    if (filter.isNotEmpty && filter != 'none') kv('filter', filter);
    return b.toString();
  }

  static List<EditRegion> _decodeRegions(String v) {
    final out = <EditRegion>[];
    for (final item in v.split('|')) {
      if (item.isEmpty) continue;
      final cs = item.split(',');
      if (cs.length < 3) continue;
      out.add(EditRegion(
        x: _num(cs[0], 0.5),
        y: _num(cs[1], 0.5),
        r: _num(cs[2], 0.04),
      ));
    }
    return out;
  }

  /// Parse the `key=value;` grammar (unknown keys ignored; malformed numbers
  /// fall back to the field default).
  static EditSpec decode(String s) {
    final e = EditSpec();
    for (final tok in s.split(';')) {
      if (tok.isEmpty) continue;
      final eq = tok.indexOf('=');
      if (eq < 0) continue;
      final k = tok.substring(0, eq);
      final v = tok.substring(eq + 1);
      double d(double fb) => _num(v, fb);
      switch (k) {
        case 'rot':
          e.rot90 = ((int.tryParse(v) ?? 0) % 4 + 4) % 4;
          break;
        case 'fliph':
          e.flipH = v == '1' || v == 'true';
          break;
        case 'flipv':
          e.flipV = v == '1' || v == 'true';
          break;
        case 'straighten':
          e.straighten = d(0);
          break;
        case 'crop':
          // Mirror native's index-walk: tolerate <4 fields, keeping each
          // field's own default (l,t=0; w,h=1) for anything missing/malformed.
          final parts = v.split(',');
          final c = <double>[0, 0, 1, 1];
          for (var i = 0; i < 4 && i < parts.length; i++) {
            c[i] = _num(parts[i], c[i]);
          }
          e.cropL = c[0];
          e.cropT = c[1];
          e.cropW = c[2];
          e.cropH = c[3];
          break;
        case 'exposure':
          e.exposure = d(0);
          break;
        case 'contrast':
          e.contrast = d(0);
          break;
        case 'highlights':
          e.highlights = d(0);
          break;
        case 'shadows':
          e.shadows = d(0);
          break;
        case 'whites':
          e.whites = d(0);
          break;
        case 'blacks':
          e.blacks = d(0);
          break;
        case 'clarity':
          e.clarity = d(0);
          break;
        case 'dehaze':
          e.dehaze = d(0);
          break;
        case 'temp':
          e.temperature = d(0);
          break;
        case 'tint':
          e.tint = d(0);
          break;
        case 'vibrance':
          e.vibrance = d(0);
          break;
        case 'saturation':
          e.saturation = d(0);
          break;
        case 'sharpness':
          e.sharpness = d(0);
          break;
        case 'noise':
          e.noise = d(0);
          break;
        case 'vignette':
          e.vignette = d(0);
          break;
        case 'autofix':
          e.autoFix = v == '1' || v == 'true';
          break;
        case 'curves':
          e.curve = [];
          for (final pt in v.split('|')) {
            if (pt.isEmpty) continue;
            final cs = pt.split(',');
            if (cs.length >= 2) {
              e.curve.add(Offset(_num(cs[0], 0), _num(cs[1], 0)));
            }
          }
          break;
        case 'text':
          e.texts = [];
          for (final item in v.split('|')) {
            if (item.isEmpty) continue;
            final parts = item.split(',');
            if (parts.length < 5) continue;
            e.texts.add(TextOverlay(
              x: _num(parts[0], 0.5),
              y: _num(parts[1], 0.5),
              size: _num(parts[2], 0.06),
              color: int.tryParse(parts[3], radix: 16) ?? 0xFFFFFF,
              // The escaped text has no raw comma, but rejoin defensively.
              text: _unesc(parts.sublist(4).join(',')),
            ));
          }
          break;
        case 'redeye':
          e.redeye = _decodeRegions(v);
          break;
        case 'heal':
          e.heal = _decodeRegions(v);
          break;
        case 'filter':
          e.filter = v;
          break;
      }
    }
    return e;
  }

  /// Net clockwise quarter-turns including flips, useful for UI badges.
  int get netRotation => rot90 % 4;

  /// A rough "is this a big geometry change" flag for preview sizing.
  double get cropZoom =>
      cropW <= 0 ? 1 : 1 / math.min(cropW, cropH).clamp(1e-3, 1.0);
}
